src/node_crypto.cc

```diff
@@ -3174,26 +3231,54 @@ void CipherBase::SetAuthTag(const FunctionCallbackInfo<Value>& args) {
 }
 
 
-bool CipherBase::SetAAD(const char* data, unsigned int len) {
+bool CipherBase::SetAAD(const char* data, unsigned int len, int plaintext_len) {
   if (ctx_ == nullptr || !IsAuthenticatedMode())
     return false;
+
   int outlen;
-  if (!EVP_CipherUpdate(ctx_,
-                        nullptr,
-                        &outlen,
-                        reinterpret_cast<const unsigned char*>(data),
-                        len)) {
-    return false;
+  const int mode = EVP_CIPHER_CTX_mode(ctx_);
+
+  // When in CCM mode, we need to set the authentication tag and the plaintext
+  // length in advance.
+  if (mode == EVP_CIPH_CCM_MODE) {
+    if (plaintext_len < 0) {
```

Comment:
As I wrote in the comment in https://github.com/nodejs/node/pull/18138#issuecomment-363339536, it should be check the plaintext length to meet nonce size in here and `Update`.

Reply:
I assume this cannot be properly detected using OpenSSL APIs so the check should be implemented within node, right?

Reply:
That is right. OpenSSL does not have the API to get the nonce size. 
We can keep it with the same way as `auth_tag_len_` and checking `plaintext_len` with it.